### PR TITLE
Asynchronous output.

### PR DESCRIPTION
**Pull request by dr-robertk**
This PR adds the possibility to spawn an extra thread for writing the output. I tested with Norne and SPE9 and the results look the same. However, it should be discussed if the cxx-11 feature std::async can be used or a similar implementation with OpenMP needs to be done (by somebody else). Right now the old state remains default, i.e. no async output.

### CODE REVIEW

**Comment by alfbr** [comment]
Actually, I was not aware of the async option, it looks like exactly what we need. Hence, I suggest we go with this. Have you checked the performance gain, e.g., do we get the same performance as disabling output? Have you tested with MPI, i.e., does the code still run, and is the thread spawned only for rank 0? Any reason why the old state should remain default?

**Comment by dr-robertk** [comment]
@alfbr: I'm still testing. With the option enabled, I gain roughly 100sec on Norne, which brings it down to 2.15 factor. All numbers are preliminary. Further testing by other people needed.

**Comment by alfbr** [comment]
Sounds like the same performance as disabling output. I can give it a go at my end.

**Comment by bska** [comment]
> it looks like exactly what we need

"Looks like" being the operative term.  Function `std::async()` is usually better than using `std::thread` directly and does go some way towards supporting task-based parallelism but it doesn't go all the way.  There is, for instance, nothing that prevents an implementation from running a `std::async()` request on the current thread (explicit execution policies help) and if it **is** run on a separate thread, you often end up creating a new, independent thread each time rather than reusing existing threads in a thread pool which is troublesome on systems for which thread creation is expensive.

It all depends on the quality of the implementation so while the interface is portable, the performance is not although implementations certainly mature and improve greatly in this area with recent iterations of compilers and runtime systems.

**Comment by atgeirr** [comment]
The code looks good to me, I don't know why Travis fails, after all we merged the necessary opm-core PR.

We should prefer std::async() over OpenMP. I must admit I am not sure about the gritty details of using async(), in particular if the std::future returned will block in its destructor. See the Note on this page for example:

http://en.cppreference.com/w/cpp/thread/async

Now I see your results, they indicate that clearly it is not blocking, and working as expected. I wonder if we might see improvements from ensuring that the std::move is efficient though (we'll probably just copy in this case since the various state classes do not have move ops).

Finally: just before Easter I discovered that on my system at least, almost all output time was spent in the RFT writing, and communicated this to @joakim-hove. I see improvements are underway here too, this may further reduce output time, and at the very least push the time when we *must* consider parallel output formats further into the future.

**Comment by dr-robertk** [comment]
I want to double check one last thing. It's a bit unclear what happens if one has spawned two threads and still we have to make sure that the execution is in the order of how threads were spawned.

**Comment by dr-robertk** [comment]
@atgeirr & @bska: I'm not how we can make sure that each time step is written in the correct order? Do you have an idea?

**Comment by atgeirr** [comment]
My simplest idea: use a `future<void>` data member, before calling async() call wait() on the future (assuming a default-constructed future will not wait), then move the result of async() into the future.
If this works it would not allow you to "stack" multiple timesteps' output calls, but I think it could work.

Alternative: create a long-lived thread, containing a `vector<WriterCall>`, push one object every timestep (insert details as needed... a bit more complex than the first idea I bet).

**Comment by dr-robertk** [comment]
@alfbr: If I disable output I get a bit slower run time (slower than with async). This is a bit strange.

**Comment by dr-robertk** [comment]
@atgeirr: I tried the future with waiting but I'm not sure If I implemented it correctly. I'll add a version.

**Comment by dr-robertk** [comment]
@atgeirr: I have added a future to ensure that the previous thread is finished. Could you take a look at this?

File: opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp [hunk 1]
```diff
@@ -264,7 +305,7 @@ namespace Opm
             vtkWriter_->writeTimeStep( timer, localState, localWellState, false );
         }
 
-        bool isIORank = true ;
+        bool isIORank = output_ ;
```
Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp]: This looks like an unrelated bugfix? Good catch.

File: opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp [hunk 2]
```diff
@@ -252,6 +254,45 @@ namespace Opm
...
 
+    struct WriterCall
+    {
+        BlackoilOutputWriter& writer_;
+        std::unique_ptr< SimulatorTimerInterface > timer_;
+        const SimulatorState state_;
+        const WellState wellState_;
+        const bool substep_;
+
+        std::future< bool > asyncWait_;
```
Comment (atgeirr) [inline opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp]: I admit this is too complicated for me to comment on quickly, my idea was adding the future to the BlackoilOutputWriter class, not the WriterCall (I guess I was unclear). Did you try that? I imagine it would be simpler.

File: opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp (hunk 1 above)
Comment (dr-robertk) [inline opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp]: Yes, I also was surprised by this. I think now it should be correct.

File: opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp (hunk 2 above)
Comment (dr-robertk) [inline opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp]: I did it this way because then the main thread does not have to wait on the output. It will simply spawn threads that wait if the previous thread was not finished. I have tried the other version too but it did not seem to give any performance improvement. I'll report back when I have some results.

**Comment by dr-robertk** [comment]
With this async wait the results are comparable to the timings without output. 

@alfbr: It would be interesting to see if you could realize the performance improvement on your side.

**Comment by alfbr** [comment]
Sure, will do. Expect results tomorrow :-)

**Comment by blattms** [comment]
FY: due to merged PR OPM/opm-core#962 this does not compile anymore.

**Comment by alfbr** [comment]
Indeed, shall I wait for a fix or attempt cherry picking revisions for repos?

**Comment by joakim-hove** [comment]
Please  wait - we are DESPERATELY trying to make progress on the output.
30. mars 2016 10:22 skrev "Alf Birger Rustad" <notifications@github.com>
følgende:

**Comment by dr-robertk** [comment]
The changes from OPM/opm-core#962 have now been included in this PR.

**Comment by alfbr** [comment]
Thank you! Finally got the first test results, and they are positive. Not by any means ideal benchmarking conditions, but it seems I get the same performance from this patch as I get when disabling output, full benefit in other words :-)

Hence, I suggest the async option is made default. We may want to keep the option of turning it off to enable benchmarking of the output writing itself. I will test with MPI too, so please hold off merging a little while yet.

**Comment by alfbr** [comment]
Hm, it seems a number of output files are missing when using the async option. The INIT and EGRID files are present, but the RFT, SMSPEC, UNRST and UNSMRY are all missing.

**Comment by dr-robertk** [comment]
@alfbr: I'll investigate this.

**Comment by atgeirr** [comment]
> Please  wait - we are DESPERATELY trying to make progress on the output

I'll listen to @joakim-hove on this, and intend to merge the opm-output changes before this. As a consequence this PR must be made against opm-output, I hope you are ok with that @dr-robertk.
